scipy.stats: generalized normal distribution

jonycgn opened the pull request:
hi,

first time contributing to scipy for me, so please don't slap me for breaking any conventions. i'm trying to follow the rules.

this is an implementation of the generalized normal distribution for scipy.stats. this distribution is widely used in signal processing and in neuroscience.

the other distribution is the same, except that the lower half has been cut off. if i find the time, i might work out entropy/stats for it, but for now this is what i have.

i've done some manual testing. how do i include this in the unit tests/run the unit tests? also, i'd be glad to receive input/comments.

thanks!

ev-br commented:
Thanks for the contribution!
Looks reasonable at first sight.

A couple of things to do here:

1. Tests: Have a look at a recent https://github.com/scipy/scipy/pull/4533 on how to add this to the standart test loop --- you want to add an entry to ` _distr_param.py`. 
2. Some more tests, if desired, go to tests/test_distributions.py. I'd add simple smoke tests comparing beta=1,2 to normal and exponential distributions. 
3. It's a good idea to make sure  tests pass locally: see http://docs.scipy.org/doc/scipy/reference/hacking.html
4. Documentation: please add an explicit formula for the PDF to the docstring. It could be a good idea to model the docstring after docstrings of other distributions, including the magic str substitution keys, %(before_notes)s and all. 
5. Vanity: please add yourself to Thanks.text

jonycgn commented:
okay, added all of this, except additional tests. i haven't managed to run tests locally yet, and since i would probably have to play around to pick error bounds, it feels like i should leave it for now.

jonycgn commented:
what do the standard distribution tests cover? i'm not sure i understand the structure of the distribution test suite.

josef-pkt commented:
The standard tests are consistency checks across the methods. For example moments, mean, variance are verified by numerical  integration with respect to the pdf. rvs is tested with kolmogorov smirnov test against the cdf, which has however relatively low power and only finds big discrepancies.

ev-br commented:
In addition to what Josef said: https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_continuous_basic.py#L91
is the main test loop over the distributions. Specific values of the shape parameters are taken from _distr_param.py.

ev-br commented:
Additional distribution is specific tests should go to https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py
roughly speaking, it's a test class per distribution. You can have a look at for example https://github.com/scipy/scipy/blob/master/scipy/stats/tests/test_distributions.py#L505 which checks that genpareto agrees with the exponential distribution for a specific value of the shape parameter.

jonycgn commented:
thanks for the pointers! i'll look at that when i find the time, and meanwhile wait for the standard tests to fail ;)

on a side note, regarding comparing to the normal/laplacian, i did some manual tests earlier and found the laplacian sf to be inaccurate. see my comment here:
https://github.com/scipy/scipy/issues/2509#issuecomment-77610250
so this would probably have to be fixed too to make that test pass ...

ev-br commented:
For the Laplace: I think it'd be best to fix it in a separate PR (I, for one, am/was certainly guilty of trying to squeeze unrelated things to a single PR, best don't do it).

Ideally, start that other PR (in a separate branch!) with a test which fails, this should go to test_distributions.py. Then fix the math like you suggested in that other issue, and send the PR :-). This way it's focused, it's easy to review, it's a clear improvement which is always welcome!

And while you're at it, have a look at the suggested format for the commit messages in the hacking.rst document I linked above: if needed, we'll help you rewrite those commit messages, but it's easier and quicker if they are already OK :-).

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4171,6 +4171,121 @@ def _entropy(self, c):
+    Notes
+    -----
+    The probability density function for `gennorm` is [1]::
+
```
ReST syntax: you need a trailing underscore for the reference, like so: [1]_ 
Or just omit it here (but keep it below).

jonycgn replied:
okay, will fix

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4171,6 +4171,121 @@ def _entropy(self, c):
+
+    `gennorm` takes ``beta`` as a shape parameter.
+    For beta = 1, it is identical to a Laplace distribution.
+    For beta = 2, it is identical to a normal distribution.
```
Not really: scipy's normal is defined with `exp(-x**2 / 2). Which corresponds to alpha=2 in the Wikipedia definition. Either use alpha=2 or ".. identical to a normal distribution with ``scale=2``." Any choice needs to be supplemented with a test asserting the equivalence in test_distributions.py

jonycgn replied:
i know about the scaling issue. this is why it says it is identical to *a* normal distribution, not to *the* normal distribution ;)

some other distributions in scipy aren't normalized to unit variance with scale=1. hence, i saw no need in standardizing this one, and decided to stick with the more concise formulation.

will make the documentation more precise to say what choice of beta and scale corresponds to which Laplace and normal distributions.

jonycgn replied:
for the test, what do we compare? the pdfs? the cdfs? everything?

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4171,6 +4171,121 @@ def _entropy(self, c):
+        return exp(self._logpdf(x, beta))
+
+    def _logpdf(self, x, beta):
+        return log(.5 * beta) - special.gammaln(1. / beta) - abs(x) ** beta
```
Style nitpick: `abs(x)**beta`, without spaces

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4171,6 +4171,121 @@ def _entropy(self, c):
+                                    gamma(1/beta)
+
+    `gennorm` takes ``beta`` as a shape parameter.
+    For beta = 1, it is identical to an exponential distribution.
```
For beta=2, it is identical to ` halfnorm` with scale=2.

jonycgn commented:
Thanks for your comments, I will work on them.

With regards to the test, I suspect the failed test (numerical derivative of cdf == pdf) is because with the parameter i set in _distr_param.py, the pdf has a kink at the median, which is where it gets evaluated.

i would like to run these tests locally to verify, but i can't compile scipy. it's missing some linear algebra dependencies. this is going to take a little longer …

jonycgn commented:
done. the kink did turn out to be the problem. added further tests and fixed all the other points, too. i ran the unit tests locally, expecting that they will pass.

the only remaining thing is the commit messages, i couldn't find any documentation about that (in the hacking.rst).

ev-br commented on scipy/stats/_continuous_distns.py:
```diff
@@ -4171,6 +4171,126 @@ def _entropy(self, c):
+
+    `gennorm` takes ``beta`` as a shape parameter.
+    For beta = 1, it is identical to a Laplace distribution.
+    For beta = 2, it is identical to a normal distribution (with scale=1/sqrt(2)).
```
Minus sign is missing in the exponential, ditto in halfgennorm.

ev-br replied:
Also please use double backticks for the code-like markup: ``scale = 1/ sqrt(2)``.

ev-br commented:
I'm sure the moments of halfgennorm could be worked out, but this is definitely not required, not in this PR anyway. Looks quite good to me as is. I made a couple of small comments about the docs, could you address those.

Commit messages. the recommended format is listed here: ishttp://docs.scipy.org/doc/numpy/dev/gitwash/development_workflow.html

This also needs a rebase. Which would be a good opportunity to squash commits and rewrite the commit message. And this will hopefully trigger the TravisCI bot, not sure why it did notbrun this time.

jonycgn commented:
i see. i'm having trouble working out how to squash the commits and do the rebase. my local repository is now messed up. i don't work with git very often. is there a way for you to help me with this? i would prefer to work things out and learn something about git, but i don't have the time to do that right now. if all else fails, i can send you a diff by email :)

ev-br commented:
Recovering from a mess can be easier than it seems to be:

```
$ git rebase --abort       # in case there is a botched rebase in progress
No rebase in progress? 

# use the name of your branch instead of pr/4604
$ git checkout -b pr/4604-save   # so that you have a backup
$ git checkout pr/4604       # switch back

# This assumes the main scipy repo is called "upstream".

$git rebase upstream/master

# .... there's a trivial merge conflict in THANKS.txt, fix it in the editor
# (search for <<<< characters)

# Now on to squashing the commits: this opens your editor. Mark unnecessary
# commits as 's' instead of 'pick'; save and close the editor; 
# the new editor window opens, write the new commit message, save and close
# the editor.

$ git rebase -i upstream/master

# check the result
$ git log

# rerun the tests
$ python runtests.py - s stats

# push the result (assuming your scipy fork is called 'origin')
$ git push origin <branch_name>

# yeah, this failed, check that the name of the branch is right... repeat with
# the -f flag (force-push)
$ git push origin <branch_name> -f

# check your PR on github, it's updated, hakuna matata
```

If something goes haywire, switch to backup, delete the original branch, clone the backup:
```
$ git checkout <backup>
$ git branch -D <PR branch>
$ git checkout -b <PR branch>  # you now have both your backup and it's copy with the name of the orig. branch
```

jonycgn commented:
okay, i ended up doing things in a slightly less elegant manner, but your recipe helped a lot. thanks for the patience. let's see what travis says...

ev-br commented:
A branch with small style tweaks: https://github.com/scipy/scipy/compare/master...ev-br:pr/4604?expand=1

Can you take over commit https://github.com/scipy/scipy/commit/738b6a5c8ba0bab7abf82f4b6fced580a61ff696

ev-br commented:
LGTM

jonycgn commented:
do i have to do anything?

ev-br commented:
I think it's ready. We typically leave things open for a while to let other people a chance to have a look if they want to. If there are no further comments for a reasonable amount of time, we'll take it as a silent yes and merge it.

gmcastil commented:
Sweet.  Thanks for helping me out.

rgommers commented:
LGTM

rgommers commented:
OK rebased and merged in 11a0771. Thanks @jonycgn, @ev-br!
